website/project/model.py

```diff
@@ -2620,3 +2769,226 @@ def to_json(self):
                       for x in self.nodes if not x.is_deleted],
             "anonymous": self.anonymous
         }
+
+
+def validate_retraction_state(value):
+    acceptable_states = [Retraction.PENDING, Retraction.RETRACTED, Retraction.CANCELLED]
+    if value not in acceptable_states:
+        raise ValidationValueError('Invalid retraction state assignment.')
+
+    return True
+
+
+class Retraction(StoredObject):
+    """Retraction object for public registrations."""
+
+    PENDING = 'pending'
+    RETRACTED = 'retracted'
+    CANCELLED = 'cancelled'
+
+    _id = fields.StringField(primary=True, default=lambda: str(ObjectId()))
+    justification = fields.StringField(default=None, validate=MaxLengthValidator(2048))
+    initiation_date = fields.DateTimeField(auto_now_add=datetime.datetime.utcnow)
+    initiated_by = fields.ForeignField('user', backref='retracted')
+    # Expanded: Dictionary field mapping admin IDs their approval status and relevant tokens:
+    # {
+    #   'b3k97': {
+    #     'has_approved': False,
+    #     'approval_token': 'Cru7wj1Puf7DENUPFPnXSwa1rf3xPN',
+    #     'disapproval_token': 'UotzClTFOic2PYxHDStby94bCQMwJy'}
+    # }
+    approval_state = fields.DictionaryField()
+    # One of 'pending', 'retracted', or 'cancelled'
+    state = fields.StringField(default='pending', validate=validate_retraction_state)
+
+    def __repr__(self):
+        parent_registration = Node.find_one(Q('retraction', 'eq', self))
+        return ('<Retraction(parent_registration={0}, initiated_by={1}) '
+                'with _id {2}>').format(
+            parent_registration,
+            self.initiated_by,
+            self._id
+        )
+
+    @property
+    def is_retracted(self):
+        return self.state == self.RETRACTED
+
+    @property
+    def pending_retraction(self):
+        return self.state == self.PENDING
+
+    def disapprove_retraction(self, user, token):
+        """Cancels retraction if user is admin and token verifies."""
+        try:
+            if self.approval_state[user._id]['disapproval_token'] != token:
+                raise InvalidRetractionDisapprovalToken('Invalid retraction disapproval token provided.')
+        except KeyError:
+            raise PermissionsError('User must be an admin to disapprove retraction of a registration.')
+
+        self.state = self.CANCELLED
+        parent_registration = Node.find_one(Q('retraction', 'eq', self))
+        parent_registration.add_log(
+            action=NodeLog.RETRACTION_CANCELLED,
+            params={
+                'node': parent_registration._id,
+                'retraction_id': self._id,
+            },
+            auth=Auth(user),
+            log_date=datetime.datetime.utcnow(),
+            save=True,
+        )
+
+    def approve_retraction(self, user, token):
+        """Add user to approval list if user is admin and token verifies."""
+        try:
+            if self.approval_state[user._id]['approval_token'] != token:
```

Comment:
Are all the admins' user IDs guaranteed to be in `self.approval_state`? If so, how is this ensured? If not, then a `PermissionsError` might be raised incorrectly.